Installation issue: libffi on power 8

Trying to build `libffi@3.3` on `power8` using `gcc@7.3.0` results in the error shown below. Using `spack@develop` checked out at commit `7813e4146459` from 12-May-2020. Even when target=ppc64le is specified:

```
$> spack install libffi target=ppc64le
...
==> Error: Failed to install libffi due to ChildError: ProcessError: Command exited with status 2:
    'make' '-j16'
7 errors found in build log:
     260    ../src/java_raw_api.c:337:1: note: declared here
     261     ffi_prep_java_raw_closure_loc (ffi_java_raw_closure* cl,
     262     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     263    libtool: compile:  /opt/spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I. -I.. -I. -I../include -Iinclude -I../src -I. -I../include -Iinclude -I../
            src -MT src/powerpc/linux64.lo -MD -MP -MF src/powerpc/.deps/linux64.Tpo -c ../src/powerpc/linux64.S -o src/powerpc/linux64.o >/dev/null 2>&1
     264    libtool: compile:  /opt/spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I. -I.. -I. -I../include -Iinclude -I../src -O3 -fomit-frame-pointer -fstric
            t-aliasing -ffast-math -mcpu=power8 -Wall -fexceptions -MT src/types.lo -MD -MP -MF src/.deps/types.Tpo -c ../src/types.c -o src/types.o >/dev/nu
            ll 2>&1
     265    In file included from ../src/powerpc/ffi.c:33:0:
  >> 266    ../src/powerpc/ffi_powerpc.h:61:9: error: '_Float128' is not supported on this target
     267     typedef _Float128 float128;
     268             ^~~~~~~~~
     269    libtool: compile:  /opt/spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I. -I.. -I. -I../include -Iinclude -I../src -O3 -fomit-frame-pointer -fstric
            t-aliasing -ffast-math -mcpu=power8 -Wall -fexceptions -MT src/powerpc/ffi_sysv.lo -MD -MP -MF src/powerpc/.deps/ffi_sysv.Tpo -c ../src/powerpc/f
            fi_sysv.c -o src/powerpc/ffi_sysv.o >/dev/null 2>&1
     270    Makefile:1273: recipe for target 'src/powerpc/ffi.lo' failed
  >> 271    make[3]: *** [src/powerpc/ffi.lo] Error 1
     272    make[3]: *** Waiting for unfinished jobs....
     273    In file included from ../src/powerpc/ffi_linux64.c:35:0:
  >> 274    ../src/powerpc/ffi_powerpc.h:61:9: error: '_Float128' is not supported on this target
     275     typedef _Float128 float128;
     276             ^~~~~~~~~
     277    Makefile:1273: recipe for target 'src/powerpc/ffi_linux64.lo' failed
  >> 278    make[3]: *** [src/powerpc/ffi_linux64.lo] Error 1
     279    libtool: compile:  /opt/spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I. -I.. -I. -I../include -Iinclude -I../src -O3 -fomit-frame-pointer -fstric
            t-aliasing -ffast-math -mcpu=power8 -Wall -fexceptions -MT src/java_raw_api.lo -MD -MP -MF src/.deps/java_raw_api.Tpo -c ../src/java_raw_api.c -o
             src/java_raw_api.o >/dev/null 2>&1
     280    libtool: compile:  /opt/spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I. -I.. -I. -I../include -Iinclude -I../src -O3 -fomit-frame-pointer -fstric
            t-aliasing -ffast-math -mcpu=power8 -Wall -fexceptions -MT src/raw_api.lo -MD -MP -MF src/.deps/raw_api.Tpo -c ../src/raw_api.c -o src/raw_api.o
            >/dev/null 2>&1
     281    libtool: compile:  /opt/spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I. -I.. -I. -I../include -Iinclude -I../src -O3 -fomit-frame-pointer -fstric
            t-aliasing -ffast-math -mcpu=power8 -Wall -fexceptions -MT src/prep_cif.lo -MD -MP -MF src/.deps/prep_cif.Tpo -c ../src/prep_cif.c -o src/prep_ci
            f.o >/dev/null 2>&1
     282    libtool: compile:  /opt/spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I. -I.. -I. -I../include -Iinclude -I../src -O3 -fomit-frame-pointer -fstric
            t-aliasing -ffast-math -mcpu=power8 -Wall -fexceptions -MT src/closures.lo -MD -MP -MF src/.deps/closures.Tpo -c ../src/closures.c -o src/closure
            s.o >/dev/null 2>&1
     283    make[3]: Leaving directory '/tmp/root/spack-stage/spack-stage-libffi-3.3-hjmp26qlv2hwyz7q4agyxsahgovsvzdt/spack-src/powerpc64le-unknown-linux-gnu
            '
     284    Makefile:1347: recipe for target 'all-recursive' failed
  >> 285    make[2]: *** [all-recursive] Error 1
     286    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-libffi-3.3-hjmp26qlv2hwyz7q4agyxsahgovsvzdt/spack-src/powerpc64le-unknown-linux-gnu
            '
     287    Makefile:622: recipe for target 'all' failed
  >> 288    make[1]: *** [all] Error 2
     289    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-libffi-3.3-hjmp26qlv2hwyz7q4agyxsahgovsvzdt/spack-src/powerpc64le-unknown-linux-gnu
            '
     290    Makefile:592: recipe for target 'all-all' failed
  >> 291    make: *** [all-all] Error 2
```

@michaelkuhn @ax3l @skosukhin

Unfortunately, no, it does not work. It's interesting, because I get the exact same error, which indicates that `defined(__FLOAT128_TYPE__)` is evaluating true in the preprocessor conditional:

<b>ffi_powerpc.h</b>
```
#if defined(__FLOAT128_TYPE__)
typedef _Float128 float128;
#elif defined(__FLOAT128__)
typedef __float128 float128;
#else
//typedef __int128 float128;
typedef char float128[16] __attribute__((aligned(16)));
#endif
```

Even with this change, the same error is returned:
```
$> spack install libffi target=ppc64le
...
  >> 247    ../src/powerpc/ffi_powerpc.h:61:9: error: '_Float128' is not supported on this
            target
     248     typedef _Float128 float128;
```

Any ideas why it thinks float128 is available?

@ax3l

--- a/var/spack/repos/builtin/packages/libffi/package.py
+++ b/var/spack/repos/builtin/packages/libffi/package.py
@@ -18,6 +18,8 @@ class Libffi(AutotoolsPackage, SourcewarePackage):
     version('3.2.1', sha256='d06ebb8e1d9a22d19e38d63fdb83954253f39bedc5d46232a05645685722ca37')
 
     patch('clang-powerpc-3.2.1.patch', when='@3.2.1%clang platform=linux')
+    # ref.: https://github.com/libffi/libffi/pull/561
+    patch('powerpc-3.3.patch', when='@3.3')
 
     @property
     def headers(self):

--- a/var/spack/repos/builtin/packages/libffi/powerpc-3.3.patch
+++ b/var/spack/repos/builtin/packages/libffi/powerpc-3.3.patch
@@ -1,0 +1,58 @@
+From de93adfb6f48100946bba2c3abad2a77a0cfde0b Mon Sep 17 00:00:00 2001
+From: Fabrice Fontaine <fontaine.fabrice@gmail.com>
+Date: Sun, 24 Nov 2019 09:52:01 +0100
+Subject: [PATCH] ffi_powerpc.h: fix build failure with powerpc7
+
+This is a patch pulled down from the following:
+https://github.com/buildroot/buildroot/blob/78926f610b1411b03464152472fd430012deb9ac/package/libffi/0004-ffi_powerpc.h-fix-build-failure-with-powerpc7.patch
+
+This issue is being hit on OpenBMC code when pulling the latest
+libffi tag and building on a P8 ppc64le machine. I verified this
+patch fixes the issue we are seeing.
+
+Below is the original commit message:
+
+Sicne commit 73dd43afc8a447ba98ea02e9aad4c6898dc77fb0, build on powerpc7
+fails on:
+
+In file included from ../src/powerpc/ffi.c:33:0:
+../src/powerpc/ffi_powerpc.h:61:9: error: '_Float128' is not supported on this target
+ typedef _Float128 float128;
+         ^~~~~~~~~
+
+Fix this build failure by checking for __HAVE_FLOAT128 before using
+_Float128, as _Float128 is enabled only on specific conditions, see
+output/host/powerpc64-buildroot-linux-gnu/sysroot/usr/include/bits/floatn.h:
+
+ /* Defined to 1 if the current compiler invocation provides a
+    floating-point type with the IEEE 754 binary128 format, and this glibc
+    includes corresponding *f128 interfaces for it.  */
+ #if defined _ARCH_PWR8 && defined __LITTLE_ENDIAN__ && (_CALL_ELF == 2) \
+     && defined __FLOAT128__ && !defined __NO_LONG_DOUBLE_MATH
+ # define __HAVE_FLOAT128 1
+ #else
+ # define __HAVE_FLOAT128 0
+ #endif
+
+Fixes:
+ - http://autobuild.buildroot.org/results/5c9dd8fb3b6a128882b6250f197c80232d8a3b53
+
+Signed-off-by: Fabrice Fontaine <fontaine.fabrice@gmail.com>
+Signed-off-by: Andrew Geissler <geissonator@yahoo.com>
+---
+ src/powerpc/ffi_powerpc.h | 2 +-
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+diff --git a/src/powerpc/ffi_powerpc.h b/src/powerpc/ffi_powerpc.h
+index 8e2f2f0e..960a5c42 100644
+--- a/src/powerpc/ffi_powerpc.h
++++ b/src/powerpc/ffi_powerpc.h
+@@ -57,7 +57,7 @@ typedef union
+   double d;
+ } ffi_dblfl;
+ 
+-#if defined(__FLOAT128_TYPE__)
++#if defined(__FLOAT128_TYPE__) && defined(__HAVE_FLOAT128)
+ typedef _Float128 float128;
+ #elif defined(__FLOAT128__)
+ typedef __float128 float128;

